### DISCUSSION TITLE
dd: failed to open ... Invalid argument

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#34] [Q&A] [answered]
I have been trying to use sdm in order to see if it would be as useful as it appears.  I am stumbling when trying to burn my first pretty standard image.  Receive the error message:
`dd: failed to open '2022-09-22-raspios-bullseye-armhf.img': Invalid argument`

I am running sdm on a laptop running Linux Mint (20.3 based on Ubuntu 20.04).  I was able to perform a couple of basic, initial configurations successfully.  In the course of trying to sort things out, I moved the .img file from the sub-directory I had it in originally to my home directory and shortened the name.
```
doug@dv6-Notebook-PC:~$ ls -l *.img
-rw-rw-r-- 1 doug doug 4261412864 Oct  5 17:56 bullseye.img
```
Attempting to burn this image fails.
```
doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --hostname KeyLime --expand-root bullseye.img
tr: read error: No such device
* Burn 'bullseye.img' to '/dev/sdd'...
dd if=bullseye.img of=/dev/sdd status=progress bs=16M iflag=direct
dd: failed to open 'bullseye.img': Invalid argument
? Exiting due to dd error
```
Is there something blindingly obvious that I have missed?

**Comment by gitbls** [comment]
No, you're doing everything correct, as far as I can tell. The first thing to sort out is the 'tr: read error'. Does your system have /dev/random on it? that's used to generate the new disk id. 

As far as the dd error, I'm mystified.  Probably not it, but what is the result of using a full path, as in /home/user/bullseye.img?

**Reply by dshaw619** [reply]
Yes, /dev/random exists.
```
doug@dv6-Notebook-PC:~$ ls -l /dev/rand*
crw-rw-rw- 1 root root 1, 8 Oct 16 15:02 /dev/random
```
Using the full path name fails in the same way:
```
doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --hostname KeyLime --expand-root /home/doug/bullseye.img
[sudo] password for doug:             
tr: read error: No such device
* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M iflag=direct
dd: failed to open '/home/doug/bullseye.img': Invalid argument
```
I had also previously tried using "~/bullseye.img".

**Comment by gitbls** [comment]
Did a bit more looking around, and I'm wondering if  'iflag=direct' is causing this.

Since I don't have an easy way to replicate it, can you try running sdm --burn again and adding the switch `--ddsw bs=16M` to the command line, overriding the sdm default which includes iflag=direct.

**Reply by dshaw619** [reply]
This made significant progress.  Still have the 'tr: read error', but the uSD card appears to have been written.  Now I'm getting a new error saying the disk is in use before expanding the /root partition.
```
doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --ddsw bs=16M --hostname KeyLime --expand-root /home/doug/bullseye.img
tr: read error: No such device
* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M
4261412864 bytes (4.3 GB, 4.0 GiB) copied, 229 s, 18.6 MB/s
254+0 records in
254+0 records out
4261412864 bytes (4.3 GB, 4.0 GiB) copied, 245.727 s, 17.3 MB/s
> Expand Root: Expand partition from 3.7GB (4.0GiB) to 14.6GB (15.7GiB)
Checking that no-one is using this disk right now ... FAILED

This disk is currently in use - repartitioning is probably a bad idea.
Umount all file systems, and swapoff all swap partitions on this disk.
Use the --no-reread flag to suppress this check.

sfdisk: Use the --force flag to overrule all checks.
? sfdisk expand last partition failed with status 1
```

/dev/sdd* are in fact not mounted.

**Reply by dshaw619** [reply]
I tried booting the uSD card I burned above in an RPi02W.  Never shows up on the WiFi net.  I will have to carefully review all the parameters I used to customize the image and connect a console device to get more info.

**Comment by gitbls** [comment]
If the disk is in use, it sounds like there's a dangling mount or loop device hanging around. Easiest would be to reboot your Mint system to clean up anything dangling, or look for any /mnt/sdm* mounts and dismount them, and clean up the associated loop devices (either `sudo losetup -D` or look at them individually and clean up any that aren't part of your normal system operation).

Let me know how it goes.

**Comment by gitbls** [comment]
Was thinking about this last night, and wondering if the dd *invalid argument* error is due to the fact that you have a `sdm --mount` or `sdm --explore` session on that _same_ IMG open in another window. 

In either case, a clean reboot and then trying your `sdm --burn` command would be enlightening.

On the 'tr' error, what is the result of the following command (this is what sdm does with the `tr` command).
```
echo $(tr -dc 'a-f0-9' < /dev/random | dd bs=1 count=8 2>/dev/null)
```

**Reply by dshaw619** [reply]
I don't believe I had any other sdm sessions running at any time during my attempts. (only had one terminal window open).  I rebooted the Linux Mint laptop, inserted the uSD card in a USB attached card reader/writer, and unmounted /dev/sdd1 and /dev/sdd2 that automatically mounted.
```
doug@dv6-Notebook-PC:~$ ls -l /mnt/sdm*
total 0
doug@dv6-Notebook-PC:~$ echo $(tr -dc 'a-f0-9' < /dev/random | dd bs=1 count=8 2>/dev/null)
171d7ea9
```
Burned the uSD card again with the same results:
```
doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --ddsw bs=16M --hostname KeyLime --expand-root /home/doug/bullseye.img
tr: read error: No such device
* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M
4261412864 bytes (4.3 GB, 4.0 GiB) copied, 204 s, 20.8 MB/s
254+0 records in
254+0 records out
4261412864 bytes (4.3 GB, 4.0 GiB) copied, 222.154 s, 19.2 MB/s
> Expand Root: Expand partition from 3.7GB (4.0GiB) to 14.6GB (15.7GiB)
Checking that no-one is using this disk right now ... FAILED

This disk is currently in use - repartitioning is probably a bad idea.
Umount all file systems, and swapoff all swap partitions on this disk.
Use the --no-reread flag to suppress this check.

sfdisk: Use the --force flag to overrule all checks.
? sfdisk expand last partition failed with status 1
```
Tried to boot uSD card in an RPi02W and never comes up on WiFi net.  I will doublecheck the settings I used to customize the image.  Also need to sort out my USB-UART connection via PuTTY;  isn't working from this laptop for some reason.

**Comment by gitbls** [comment]
Wow, the tr issue is very strange. That line of code is right from sdm. I'm baffled as to why it would fail on Mint in a script but work from the command line. I've been looking into some tr-less replacement code, but haven't implemented anything yet.

As far as the other issue, I'm thinking that your disk automounter is the culprit.

Please have a look at the system logs and see if there's anything useful there in terms of automounter log entries. I'd expect to see some about the time you're doing the burn.

Not sure what does that on Mint (something in file explorer?), but if you can figure it out, stop it from automounting and try it again from the top, that would be VERY helpful in chasing this down.

**Reply by dshaw619** [reply]
In Linux Mint, the file explorer is Nemo 5.2.4.  I went into Edit > Preferences > Behavior > Media Handling and unchecked all options (1st 3 had been checked previously):
- Automatically mount removeable media ...
- Automatically open a folder for automounted media
- Prompt or autorun/autostart programs when media are inserted
- Automatically close the device's tab, pane, or window ...

Tried my sdm --burn command again and made more progress:
```
doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --ddsw bs=16M --hostname KeyLime --expand-root /home/doug/bullseye.img
tr: read error: No such device
* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M
4244635648 bytes (4.2 GB, 4.0 GiB) copied, 201 s, 21.1 MB/s
254+0 records in
254+0 records out
4261412864 bytes (4.3 GB, 4.0 GiB) copied, 223.467 s, 19.1 MB/s
> Expand Root: Expand partition from 3.7GB (4.0GiB) to 14.6GB (15.7GiB)
Checking that no-one is using this disk right now ... OK

Disk /dev/sdd: 14.86 GiB, 15931539456 bytes, 31116288 sectors
Disk model: MassStorageClass
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x1d124286

Old situation:

Device     Boot  Start     End Sectors  Size Id Type
/dev/sdd1         8192  532479  524288  256M  c W95 FAT32 (LBA)
/dev/sdd2       532480 8323071 7790592  3.7G 83 Linux

/dev/sdd2: 
New situation:
Disklabel type: dos
Disk identifier: 0x1d124286

Device     Boot  Start      End  Sectors  Size Id Type
/dev/sdd1         8192   532479   524288  256M  c W95 FAT32 (LBA)
/dev/sdd2       532480 31116287 30583808 14.6G 83 Linux

The partition table has been altered.
Calling ioctl() to re-read partition table.
Re-reading the partition table failed.: Device or resource busy
The kernel still uses the old table. The new table will be used at the next reboot or after you run partprobe(8) or kpartx(8).
Syncing disks.
> Set new disk ID '' on '/dev/sdd'
sfdisk: unrecognized option '--disk-id'
Try 'sfdisk --help' for more information.
? sfdisk error setting diskid 1
```

**Comment by gitbls** [comment]
Ah, perfect. It's all working now, except for the tr command, which is causing the sfdisk command to fail.
If you're up for seeing it all work quickly, you can do the following. I haven't done extensive testing on it, but it should work:

sudo edit /usr/local/sdm/sdm-cmdsubs, and find the line:
```
    newdiskid="$(tr -dc 'a-f0-9' < /dev/random | dd bs=1 count=8 2>/dev/null)"
````
Replace it with
```
    newdiskid="$(printf  "%x" "$((RANDOM*RANDOM*RANDOM*RANDOM&0xFFFFFFFF))")"
```
Save the file, and try your burn command again (try it without the `--ddsw` switch if possible)

**Reply by dshaw619** [reply]
That change to sdm-cmdsubs seems to have fixed the tr error.  

It seems that there is different behavior  with respect to the someone else is using the disk, though, depending on how I connect the uSD card and reader.  

- If I try to boot the laptop with the reader connected to the USB port and the uSD card inserted, the laptop doesn't boot (trying to boot from the uSD card?).  
- If I boot with the reader disconnected from the laptop and then connect it, I get the someone else is using the disk errors (even though I haven't opened the file explorer Nemo).
- If I boot with the reader connected but the uSD card not inserted, and then insert the uSD card after the laptop boots, I get the most successful burn.

```
doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --ddsw bs=16M --hostname KeyLime --expand-root /home/doug/bullseye.img
[sudo] password for doug:             
* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M
4244635648 bytes (4.2 GB, 4.0 GiB) copied, 200 s, 21.2 MB/s
254+0 records in
254+0 records out
4261412864 bytes (4.3 GB, 4.0 GiB) copied, 220.162 s, 19.4 MB/s
> Expand Root: Expand partition from 3.7GB (4.0GiB) to 14.6GB (15.7GiB)
Checking that no-one is using this disk right now ... OK

Disk /dev/sdd: 14.86 GiB, 15931539456 bytes, 31116288 sectors
Disk model: MassStorageClass
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x1d124286

Old situation:

Device     Boot  Start     End Sectors  Size Id Type
/dev/sdd1         8192  532479  524288  256M  c W95 FAT32 (LBA)
/dev/sdd2       532480 8323071 7790592  3.7G 83 Linux

/dev/sdd2: 
New situation:
Disklabel type: dos
Disk identifier: 0x1d124286

Device     Boot  Start      End  Sectors  Size Id Type
/dev/sdd1         8192   532479   524288  256M  c W95 FAT32 (LBA)
/dev/sdd2       532480 31116287 30583808 14.6G 83 Linux

The partition table has been altered.
Calling ioctl() to re-read partition table.
**Re-reading the partition table failed.: Device or resource busy**
The kernel still uses the old table. The new table will be used at the next reboot or after you run partprobe(8) or kpartx(8).
Syncing disks.
> Set new disk ID '3db55fd2' on '/dev/sdd'
**sfdisk: unrecognized option '--disk-id'**
Try 'sfdisk --help' for more information.
? sfdisk error setting diskid 1
```
Not sure if the "Re-reading the partition table failed.: Device or resource busy"  and "sfdisk: unrecognized option '--disk-id'" errors near the end are significant.

I haven't tried to boot this latest uSD in the RPi02W yet.  I need to step away from this for a bit and make sure I've really got the sequences of what does and doesn't work correct.

**Comment by gitbls** [comment]
Can't help you with the boot issues/questions on your laptop. I don't have Mint here, so can't look at the logs to see what's going on and offer suggestions. Sounds like you have a scenario that works? Generally, it's best to disable ALL automatic disk handling so the system doesn't do something unexpected behind your back.

I see that there's still an issue (rereading partition table failed: device or resource is busy). You didn't say which boot scenario this is, but there should be NO resources like that in a good burn.

I do observe that there is still an sfdisk issue (sfdisk unrecognized option --diskid). sfdisk on RasPiOS is "sfdisk from util-linux 2.36.1". Please let me know what `sfdisk --version` says on your system.

**Comment by gitbls** [comment]
I've nailed the sfdisk issue. Looks like Ubuntu 20.04 had an sfdisk that didn't have `--disk-id`. I've updated the code to try the "old way" (using fdisk) if sfdisk fails, so I think that's resolved.

By my tally, the remaining problem you're having is overly aggressive automatic disk mounting. Not sure what I can do about that, so I'm going to leave that with you to resolve. Let me know if I'm missing something.

I'm planning to check in these updates Tuesday morning. I'm going to be more-or-less tied up 10/19-10/27 with not much time to work on this.  So, find bugs quickly, especially if you need them fixed before my upcoming constrained time.

**Comment by gitbls** [comment]
As I suggested, update checked in this morning that will handle down-rev sfdisk. Please let me know how it goes.

**Reply by dshaw619** [reply]
Sorry it took me a while to test further and respond.  Greatly appreciate all your help with this and understand if you won't be readily available for a while or ultimately can't support sdm on Linux Mint.

After additional testing with the latest update (I simply reinstalled over the existing files using the EZsdmInstaller), some things are resolved and some not.

- tr: no such device error    resolved.
- sfdisk:  unrecognized option    resolved.  
- dd: invalid argument    --ddsw bs=16M switch on burn command needed
- disk in use    Still a problem.  Most of the time after rebooting the Linux Mint laptop and inserting the uSD card, sdm-burn fails saying "This disk is currently in use ...".  On a couple of occasions the burn seems to have completed successfully, but after the partition table was altered, I get a message "Re-reading the partition table failed.: Device or resource busy".  

I haven't been able to pin down or recreate a sequence that gets past the first disk in use error with any consistency.  

In my most recent test, I rebooted the Mint laptop with the card reader/writer plugged in but no uSD card inserted.  After the system booted, I inserted the uSD card.  I see now sign of any automount taking place.

```
doug@dv6-Notebook-PC:~$ df -h
df: /run/user/1000/doc: Operation not permitted
Filesystem           Size  Used Avail Use% Mounted on
udev                 1.9G     0  1.9G   0% /dev
tmpfs                393M  1.7M  391M   1% /run
/dev/sda5            109G   53G   52G  51% /
tmpfs                2.0G  436K  2.0G   1% /dev/shm
tmpfs                5.0M  4.0K  5.0M   1% /run/lock
tmpfs                2.0G     0  2.0G   0% /sys/fs/cgroup
/dev/sda1            511M  4.0K  511M   1% /boot/efi
tmpfs                393M   20K  393M   1% /run/user/1000
/home/doug/.Private  109G   53G   52G  51% /home/doug

doug@dv6-Notebook-PC:~$ cat /proc/mounts | grep sd
/dev/sda5 / ext4 rw,relatime,errors=remount-ro 0 0
cgroup2 /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate 0 0
/dev/sda1 /boot/efi vfat rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
gvfsd-fuse /run/user/1000/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
doug@dv6-Notebook-PC:~$ cat /var/log/syslog | grep mount
Oct 19 10:05:06 dv6-Notebook-PC systemd[1]: proc-sys-fs-binfmt_misc.automount: Got automount request for /proc/sys/fs/binfmt_misc, triggered by 790 (update-binfmts)
Oct 19 10:05:06 dv6-Notebook-PC kernel: [    4.175066] EXT4-fs (sda5): mounted filesystem with ordered data mode. Opts: (null)
Oct 19 10:05:06 dv6-Notebook-PC kernel: [    5.002739] systemd[1]: Set up automount Arbitrary Executable File Formats File System Automount Point.
Oct 19 10:05:06 dv6-Notebook-PC kernel: [    5.043075] systemd[1]: Starting Remount Root and Kernel File Systems...
Oct 19 10:05:06 dv6-Notebook-PC kernel: [    5.065859] EXT4-fs (sda5): re-mounted. Opts: errors=remount-ro
Oct 19 10:05:06 dv6-Notebook-PC kernel: [    5.070133] systemd[1]: Finished Remount Root and Kernel File Systems.
Oct 19 10:05:34 dv6-Notebook-PC dbus-daemon[1305]: [session uid=1000 pid=1305] Activating via systemd: service name='org.gtk.vfs.UDisks2VolumeMonitor' unit='gvfs-udisks2-volume-monitor.service' requested by ':1.13' (uid=1000 pid=1411 comm="csd-automount " label="unconfined")
Oct 19 10:05:34 dv6-Notebook-PC dbus-daemon[1305]: [session uid=1000 pid=1305] Activating via systemd: service name='org.gtk.vfs.GoaVolumeMonitor' unit='gvfs-goa-volume-monitor.service' requested by ':1.13' (uid=1000 pid=1411 comm="csd-automount " label="unconfined")
Oct 19 10:05:35 dv6-Notebook-PC dbus-daemon[1305]: [session uid=1000 pid=1305] Activating via systemd: service name='org.gtk.vfs.GPhoto2VolumeMonitor' unit='gvfs-gphoto2-volume-monitor.service' requested by ':1.13' (uid=1000 pid=1411 comm="csd-automount " label="unconfined")
Oct 19 10:05:35 dv6-Notebook-PC dbus-daemon[1305]: [session uid=1000 pid=1305] Activating via systemd: service name='org.gtk.vfs.MTPVolumeMonitor' unit='gvfs-mtp-volume-monitor.service' requested by ':1.13' (uid=1000 pid=1411 comm="csd-automount " label="unconfined")
Oct 19 10:05:35 dv6-Notebook-PC dbus-daemon[1305]: [session uid=1000 pid=1305] Activating via systemd: service name='org.gtk.vfs.AfcVolumeMonitor' unit='gvfs-afc-volume-monitor.service' requested by ':1.13' (uid=1000 pid=1411 comm="csd-automount " label="unconfined")
Oct 19 10:05:43 dv6-Notebook-PC systemd[1164]: run-user-116-gvfs.mount: Succeeded.
Oct 19 10:05:43 dv6-Notebook-PC systemd[1]: run-user-116-gvfs.mount: Succeeded.
Oct 19 10:05:43 dv6-Notebook-PC systemd[1268]: run-user-116-gvfs.mount: Succeeded.
Oct 19 10:05:43 dv6-Notebook-PC systemd[1268]: run-user-116.mount: Succeeded.
Oct 19 10:05:43 dv6-Notebook-PC systemd[1]: run-user-116.mount: Succeeded.
Oct 19 10:11:12 dv6-Notebook-PC sudo: pam_ecryptfs: pam_sm_authenticate: /home/doug is already mounted

doug@dv6-Notebook-PC:~$ cat /var/log/syslog | grep sdd
Oct 19 10:05:06 dv6-Notebook-PC lvm[370]:   /dev/sdd: open failed: No medium found
Oct 19 10:05:06 dv6-Notebook-PC lvm[370]:   /dev/sdd: open failed: No medium found
Oct 19 10:05:06 dv6-Notebook-PC kernel: [    3.480636] sd 6:0:0:2: [sdd] Attached SCSI removable disk
Oct 19 10:08:04 dv6-Notebook-PC kernel: [  185.866318] sd 6:0:0:2: [sdd] 31116288 512-byte logical blocks: (15.9 GB/14.8 GiB)
Oct 19 10:08:04 dv6-Notebook-PC kernel: [  185.880192]  sdd: sdd1 sdd2
Oct 19 10:14:56 dv6-Notebook-PC kernel: [  597.732987]  sdd: sdd1 sdd2
```
Excerpt from /var/log/syslog around 10:14:
```
Oct 19 10:13:05 dv6-Notebook-PC rtkit-daemon[1179]: Supervising 4 threads of 2 processes of 1 users.
Oct 19 10:14:56 dv6-Notebook-PC kernel: [  597.732987]  sdd: sdd1 sdd2
Oct 19 10:14:53 dv6-Notebook-PC rtkit-daemon[1179]: message repeated 5 times: [ Supervising 4 threads of 2 processes of 1 users.]
Oct 19 10:16:03 dv6-Notebook-PC rtkit-daemon[1179]: Supervising 4 threads of 2 processes of 1 users.
```

**Comment by gitbls** [comment]
Great that we've made some progress! You said "dd: invalid argument --ddsw bs=16M switch on burn command needed". I'm not sure how to interpret that. Can you please expand on this a bit?

As far as the disk in use problem, in case it's not clear, those messages are coming from system utilities such as parted and/or sfdisk, not from sdm itself. 

Also, one other thing to try: edit sdm-cmdsubs, and in the expandpartition function, uncomment the two lines #parted and #errifrc, and comment out the two lines following those (echo ",+" and errifrc).

If that doesn't resolve the issue, what is the result if you don't use the `--expand-partition` switch on the burn command?

**Reply by dshaw619** [reply]
Sorry I wasn't more clear.  The error from dd occurs when I use sdm-burn without the --ddsw bs=16M switch.  This was part of the original problem in this thread.  You had suggested I try using that switch which eliminated the error from dd.  When I tested the patch you suggested for the tr issue, you asked that I try without the --ddsw bs=16M switch.  Looking back through this thread I don't think I reported that I had tested without the switch and it failed.  Here is the most recent attempt:

```
	doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --hostname KeyLime --expand-root /home/doug/bullseye.img
	[sudo] password for doug:             
	* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
	dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M iflag=direct
	dd: failed to open '/home/doug/bullseye.img': Invalid argument
```
I will try your suggested edits to sdm-cmdsub or not using --expand-partition switch later this afternoon.

Thanks again for all your help.

**Comment by gitbls** [comment]
Hi, Wondering if you've had an opportunity to try not using --expand-partition or doing the edit to sdm-cmdsubs?

**Reply by dshaw619** [reply]
Sorry for the delay.  Got tied up with some other things.  I tried your suggested edit to sdm-cmdsubs and it seems to successfully burned the uSD.  Unfortunately, the RPi02W still doesn't appear on my network when booted from the uSD card, but I'll have to carefully go over all the network settings, etc. and/or hook up a console device to pursue that.
```
    parted -m /dev/$dev u s resizepart 2 $newsize
    errifrc $? "? parted resizepart failed with status"
    #echo ",+" | sfdisk -N 2 $fulldevname
    #errifrc $? "? sfdisk expand last partition failed with status"
```
```
	doug@dv6-Notebook-PC:~$ sudo /usr/local/sdm/sdm --burn /dev/sdd --ddsw bs=16M --hostname KeyLime --password-user XXX --expand-root /home/doug/bullseye.img
	[sudo] password for doug:             
	* Burn '/home/doug/bullseye.img' to '/dev/sdd'...
	dd if=/home/doug/bullseye.img of=/dev/sdd status=progress bs=16M
	4261412864 bytes (4.3 GB, 4.0 GiB) copied, 214 s, 20.0 MB/s
	254+0 records in
	254+0 records out
	4261412864 bytes (4.3 GB, 4.0 GiB) copied, 229.901 s, 18.5 MB/s
	> Expand Root: Expand partition from 3.7GB (4.0GiB) to 14.6GB (15.7GiB)
	> Set new disk ID 'eaa73540' on '/dev/sdd'
	sfdisk: unrecognized option '--disk-id'
	Try 'sfdisk --help' for more information.
	>%sfdisk failed; Trying fdisk to set Disk ID
	> Disk ID eaa73540 set successfully with fdisk
	mount: /dev/sdd2 mounted on /mnt/sdm.
	mount: /dev/sdd1 mounted on /mnt/sdm/boot.
	> Set new Disk ID 'eaa73540' in /etc/fstab and /boot/cmdline.txt
	> RasPiOS firstboot service not disabled; at least one of --expand-root and --regen-ssh-host-keys not set
	> Flush accumulated burn log messages
	> Set hostname 'KeyLime'
	> First System Boot Custom Boot Scripts disabled
	> First System Boot automatic restart enabled
	> Change passwords per --password-pi, --password-user, and/or --password-root
	> Execute burn-time password changes [/etc/sdm/assets/burn-change-passwords] in nspawn
	* Burn Completed
	* Storage '/dev/sdd' is ready
	umount: /mnt/sdm/boot unmounted
	umount: /mnt/sdm unmounted
```

**Comment by gitbls** [comment]
The best way to figure out what's happening is to look at the system logs (/var/log/syslog). That can provide a fair amount of detail as to what is happening when your Pi2w boots. You can mount the SD card on another Linux system and pull off the syslog file for a good look at it.

If you are unable to sort out what's going on, please post the entire file here so I can have a look at it. Trying to resolve this without the logs is likely to be quite unsuccessful.

**Reply by dshaw619** [reply]
Sorry I've been so slow in following up on this.  I've been trying to sort through various combinations and have finally achieved successful burns that boot and believe I've isolated the circumstances in Linux Mint that somehow were giving problems.

It seems that if I execute a df -h command on my Linux Mint laptop (which I was doing to verify that the uSD card was not mounted) and then execute sdm with the --burn command is when I get errors from the dd command in the sdm script.  If I reboot and don't execute a df -h command the sdm --burn command runs successfully and I am able to boot the uSD card in my RPi02W.

I've encountered some other oddities and questions for which I'll open new threads so this one won't get even longer and can be closed.

Thanks for all your help.

**Comment by gitbls** [comment, marked as answer]
We've determined that your version of dd isn't too old, but in one of your other threads you mentioned the kernel version on your Linux Mint, and it's pretty old. My current suspicion is that your Linux Mint OS doesn't like `iflag=direct` on disk files for some reason. 

Also, there's no way to close a discussion topic. Issues can be closed, of course.
